src/data/nat/factorization/basic.lean

```diff
@@ -237,6 +238,23 @@ end
 lemma pow_factorization_le {n : ℕ} (p : ℕ) (hn : n ≠ 0) : p ^ n.factorization p ≤ n :=
 le_of_dvd hn.bot_lt (nat.pow_factorization_dvd n p)
 
+/-- A crude upper bound on `n.factorization p` -/
+lemma factorization_lt {n : ℕ} (p : ℕ) (hn : n ≠ 0) : n.factorization p < n :=
+begin
+  by_cases pp : p.prime, swap, { simp [factorization_eq_zero_of_non_prime n pp], exact hn.bot_lt },
+  rw ←pow_lt_iff_lt_right pp.two_le,
+  apply lt_of_le_of_lt (pow_factorization_le p hn),
+  exact lt_of_lt_of_le (lt_two_pow n) (pow_le_pow_of_le_left (by linarith) pp.two_le n),
+end
+
+/-- An upper bound on `n.factorization p` -/
+lemma factorization_le_of_le_pow {n p b : ℕ} (hb : n ≤ p ^ b) : n.factorization p ≤ b :=
+begin
+  rcases eq_or_ne n 0 with rfl | hn, { simp },
+  by_cases pp : p.prime, swap, { simp [factorization_eq_zero_of_non_prime n pp] },
+  exact (pow_le_iff_le_right pp.two_le).1 (le_trans (pow_factorization_le p hn) hb),
```

Comment:
Maybe it's personal taste, but I don't really get this style. Why swap the order of two one-line proofs? I understand it (and bracketing the first but not the second) when the second is significantly longer, like in the first line here. But I'd personally prefer this to be 
```suggestion
  rcases eq_or_ne n 0 with rfl | hn, { simp },
  by_cases pp : p.prime, 
  { exact (pow_le_iff_le_right pp.two_le).1 (le_trans (pow_factorization_le p hn) hb) },
  { simp [factorization_eq_zero_of_non_prime n pp] }
```

Reply:
I feel that handling the `¬ p.prime` case like this makes it a little clearer to the reader that there's an asymmetry between the the trivial case that we dismiss quickly and the main case that we're actually interested in.

Often I'll have introduced a line like this when I've started by drafting the proof with an assumption of `p.prime` and then gone back to try weakening the assumptions.  This could instead have been spelled as `cases em' p.prime with hp hp,` but `by_cases pp : p.prime, swap,` is marginally shorter — but perhaps it's not worth the saving.  

Anyway, I'm happy to go with your rewrite here — just explaining my thinking!